### ISSUE TITLE
Have Bundler respect RubyGems' `--user-install`

### ISSUE THREAD

**Issue by duckinator** [rubygems/rubygems#5682]
### Describe the problem as clearly as you can

Bundler should respect RubyGems installation directory unless explicitly overwritten.

E.g., putting `install: --user-install` in `~/.gemrc` should affect `bundle install`, unless Bundler's install path was overridden.

I don't have a great understanding of the nuance, since this frequently gets brought up as a side-discussion for other things. I'm hoping by having an issue _specifically_ about this, we might be able to consolidate the information about it.

Part of this should include ripping out the use of `sudo` (as I mentioned in https://github.com/rubygems/rubygems/issues/4031#issuecomment-852773654 and others seemed to agree with), or limiting it to macOS. It is still unclear to me why macOS' system Ruby needs it, but if it still does, we can simply limit using `sudo` to that and nothing else.

Related issues/PRs:
- #1394 
- #5327 
- #2847 
- #4031 
- #4028

(Other sections from the issue template were removed because they're not relevant.)

**Comment by duckinator** [comment]
cc @deivid-rodriguez @indirect, since y'all previously had input on how this should be approached.

Feel free to ping anyone else you think would have some good input on this.

**Comment by deivid-rodriguez** [comment]
This sounds good to me, but regardless of `--user-install` being configured in RubyGems, I think we should also implement the same behavior in #5327, but for Bundler. Namely, if the default `GEM_HOME` is not writable, default to installing somewhere else that's writable. It could be `Gem.user_dir` like `--user-install` does, but I'd rather make it `vendor/bundle`, since that's what the `sudo` prompt currently mentions: https://github.com/rubygems/rubygems/blob/eba5fc5fccc822f7cb859c7c51f657c3ffdec507/bundler/lib/bundler.rb#L544-L555

**Comment by indirect** [comment]
Personally I'm a big fan of `.bundle` (and I actually set `bundle config path .bundle` globally myself). It keeps the gems in the application directory, but puts them somewhere that everyone is already ignoring in git.

**Comment by deivid-rodriguez** [comment]
Oh yeah, we already discussed `.bundle` vs `vendor/bundle` at https://github.com/rubygems/rubygems/issues/4534.

I'm good with `.bundle` too to fix this particular issue, and that's the current plan for Bundler 3 default anyways, so makes sense to unify that here too.

**Comment by duckinator** [comment]
I like `.bundle`, too -- I see `vendor/` used a lot actually-vendored projects, and using `.bundle` avoids trampling that.

But now I'm wondering if having Bundler default to `.bundle` all the time might be a good option. I've personally never _wanted_ Bundler to install things globally -- it's always for specific projects. It also makes the need to respect `--user-install` irrelevant, and avoids the complexity that comes with that.

Is there a downside to Bundler defaulting to `.bundle`, as long as the setting can still be overridden?

**Comment by deivid-rodriguez** [comment]
It was already decided a long time that `.bundle` will be the default for Bundler 3. But it needs to be Bundler 3 given that changing this default is not backwards compatible. For this specific case though, it seems fine to switch to `.bundle` right away.

**Comment by duckinator** [comment]
@deivid-rodriguez so, just to confirm I'm understanding correctly, you're thinking basically this, right?

1. for 2.x, switch to `.bundle` if RubyGems is using `--user-install`
2. with Bundler 3, switch to `.bundle` for everything by default

If so, I like this approach a lot.

**Comment by deivid-rodriguez** [comment]
Not exactly, my thinking is:

1. for 2.x, switch to install to `.bundle` if GEM_HOME is not writable.
2. with Bundler 3, switch to `.bundle` for everything by default.

**Comment by duckinator** [comment]
Sounds good to me. 👍

**Comment by duckinator** [comment]
I would love to get #5327 sorted out and also implement this, in the next few months:

> 1. for 2.x, switch to install to .bundle if GEM_HOME is not writable.
> 2. with Bundler 3, switch to .bundle for everything by default.

It'd be very nice to have RubyGems and Bundler, by default, switch to user-controlled install directories if there's not write permissions.

**Comment by indirect** [comment]
RubyGems without write permissions has a problem: it needs to install binstubs into your PATH. I expect that a user who runs `gem install rake` but gets a silent implied `--user-install` will think that RubyGems is broken and file an issue, because they will get `no such command: rake`, even when the gem is installed.

This is why, historically, macOS users were told to use sudo: you need root permissions to put binstubs into a directory that's already in the default PATH. At absolute minimum, we will need to check $PATH for `~/.local/share/gem/ruby/3.0.0/bin` and print a giant warning if it's not present.

Bundler, happily, does not have this problem at all, and `bundle exec` takes care of everything for us.

**Comment by indirect** [comment]
@eregon I don't think that warning is sufficient, because it doesn't explain anything about how to resolve the problem. A warning that just says "the gem you just installed won't run" is a good starting point, but we should explain _why_ the gem you just installed won't run, and how to fix it so that the gem you just installed _will_ run.

Do ruby version managers already add the user-install bindir to the PATH that's available? Can we get them to do that?

**Comment by duckinator** [comment]
@indirect yes, that problem is part of why `--user-install` isn't being chosen by default 100% of the time, and was being proposed only as a fallback. The warning also does explain why it won't run and how to fix it. The exact text from what @eregon linked to is:

> WARNING:  You don't have /home/eregon/.gem/truffleruby/20.2.0-dev-dc3a61a7/bin in your PATH,
>	  gem executables will not run.

I don't know about other Ruby version managers, but [`chruby` in particular sets `$PATH`, `$GEM_HOME`, and `$GEM_PATH`](https://github.com/postmodern/chruby#features). So `chruby` should never fall back to `--user-install`.

**Comment by duckinator** [comment]
FWIW: It may make sense to have Bundler 2.x fall back to `.bundle`, then come back to the RubyGems side of it. The Bundler side seems a lot more cut-and-dry.

We've been going in circles on the RubyGems side for literally half a decade. There's no need to hold Bundler back because of that.

**Comment by indirect** [comment]
@duckinator I recognize the people posting on this ticket do understand what to do based on that warning, but I am trying to point out that we regularly get tickets opened by people who do _not_ know what to do based on that warning. Here's an example of the kind of warning I think we should print:

> WARNING: Because <you used --user-install / RubyGems did not have permission to write to Gem.bindir>, the executable `rake` has been placed into the directory `/home/eregon/.gem/truffleruby/20.2.0-dev-dc3a61a7/bin`, which is not in your $PATH. You will not be able to run `rake` as a result. To resolve this issue, you will need to add the directory to your $PATH. For example, if you are using zsh you could run: `echo 'export PATH="$PATH:/home/eregon/.gem/truffleruby/20.2.0-dev-dc3a61a7/bin"' >> ~/.zshrc` and then open a new shell.

Having said that, my personal positions are (from least controversial to most controversial):

1) Bundler should fall back on `.bundle` if anything is unwritable
2) RubyGems should automatically fall back on the user-install directory if anything is unwritable
3) Bundler should not read a single one of RubyGems' configuration flags as a one-off
4) RubyGems should offer a `path` configuration option if you want your gems in e.g. ~/.local/gems
4) RubyGems should not have `--user-install` as a runtime flag
5) RubyGems should not apply runtime flags from a config file

I think we should absolutely implement and merge both #1 and #2 as noncontroversial pure wins. We can continue to ~argue~ discuss the rest. :)

**Comment by duckinator** [comment]
I'm okay with making the warning more detailed, if you think it'd help! We could also add e.g. `gem help user_install` or similar with a detailed explanation of how to adjust your PATH on Windows, macOS, and Linux/*BSD, and have the warning point at that.

Also: I agree with all 6 of your ideas. :3

**Comment by SamB** [comment]
> Here's an example of the kind of warning I think we should print:
> 
> > WARNING: Because <you used --user-install / RubyGems did not have permission to write to /usr/bin>, the executable `rake` has been placed into the directory `/home/eregon/.gem/truffleruby/20.2.0-dev-dc3a61a7/bin`, which is not in your $PATH. You will not be able to run `rake` as a result. To resolve this issue, you will need to add the directory to your $PATH. For example, if you are using zsh you could run: `echo 'export PATH="$PATH:/home/eregon/.gem/truffleruby/20.2.0-dev-dc3a61a7/bin"' >> ~/.zshrc` and then open a new shell.

That kind of makes sense, except that the part where the path contains `truffleruby/20.2.0-dev-dc3a61a7/bin` leads me to expect the `PATH` to end up a mile long if you do this more than once.

(I also suspect it would make sense to provide one or more suggestions that work for shells that aren't zsh: it might be a good shell, but it's not the only shell. I imagine a reference to the documentation might be in order for things that aren't compatible with /bin/sh or zsh.)

Oh, and isn't that stuff supposed to go in the shell profile, not the rc file?

**Comment by rubyFeedback** [comment]
Note that this was recently critizised here:

https://felipec.wordpress.com/2025/08/17/ruby-gems-still-broken/

I have no substantial pro or con opinion either way, but I think bundler and rubygems' behaviour should be the same when it comes to --user-install, as much as that is possible. I am a gem-user, not a bundler-guy though; to me --user-install kind of meant to e. g. install things into a home-directory. This is sometimes necessary - I remember on a university campus, the Linux machines we used did not allow us to modify anything under / or /usr so I had to use my home directory. Wheres on e. g. my laptop, I can do what I want.

> Oh, and isn't that stuff supposed to go in the shell profile, not the rc file?

Here I also have no specific opinion, but I do not think ruby or anyone else should assume much at all about rc-files. I source them very "unorthodox" for instance; ruby autogenerates various shell files, including aliases for the shell I use (bash uses another format than zsh which uses a slightly different format to fish which uses a slightly different format as cmder on windows etc...). We should not assume every computer to use or have the same layout or files / file management per se. Ruby should be as flexible as possible at all times - that mindset is in general better than assuming "xyz must be at abc but not at def".

**Comment by duckinator** [comment]
Hey, I'm sorry for the poor communication around this. I know the way RubyGems and Bundler behave out-of-the-box has been pretty rough, and we've only in the last few years really made _any_ progress.

We decided a while ago that the next major Bundler version would install to `.bundle` in the project directory by default, which means this issue's entire premise is outdated. This issue should have been closed when that decision was made.

You can look at #8650 to keep track of the progress on the next major Bundler version, which will include the change to using `.bundle`.

Given that the alternative solution we switched to is already merged and waiting for release later this year, I'm going to go ahead and close this.
